Loop/Managers/LoopDataManager.swift

```diff
@@ -94,11 +101,80 @@ class LoopDataManager {
             }
         } catch let error {
             lastLoopError = error
+            self.uploadLoopStatus(nil)
         }
 
         notify()
     }
 
+    private var lastTempBasalUploaded: DoseEntry?
+
+    func uploadLoopStatus(recommendation: TempBasalRecommendation?) {
+
+        let statusTime = NSDate()
+
+        let glucose = deviceDataManager.glucoseStore?.latestGlucose
+
+        let iob: IOBStatus?
+
+        if let iobValues = insulinOnBoardValues, iobValue = iobValues.first {
+            iob = IOBStatus( timestamp: iobValue.startDate, iob: iobValue.value)
+        } else {
+            iob = nil
+        }
+
+       let eventualBG = self.predictedGlucose?.last?.quantity.doubleValueForUnit(HKUnit.milligramsPerDeciliterUnit())
+
+        let loopSuggested: LoopSuggested?
+
+        let glucoseVal = glucose?.quantity.doubleValueForUnit(HKUnit.milligramsPerDeciliterUnit())
+
+        let predBGs: [Double]?
+        if let predicted = predictedGlucose {
+            predBGs = predicted[0...10].map({ (value) -> Double in
+                value.quantity.doubleValueForUnit(HKUnit.milligramsPerDeciliterUnit())
+            })
+        } else {
+            predBGs = nil
+        }
+
+        if let recommendation = recommendation, let glucoseVal = glucoseVal, let eventualBG = eventualBG {
+            loopSuggested = LoopSuggested(timestamp: recommendation.recommendedDate, rate: recommendation.rate, duration: recommendation.duration, eventualBG: Int(eventualBG), bg: Int(glucoseVal), correction: recommendedBolus, predBGs: predBGs)
+        } else {
+            loopSuggested = nil
+        }
+
+        let loopEnacted: LoopEnacted?
+        if let tempBasal = lastTempBasal where tempBasal.unit == .unitsPerHour &&
+            lastTempBasalUploaded?.startDate != tempBasal.startDate {
+            let duration = tempBasal.endDate.timeIntervalSinceDate(tempBasal.startDate)
+            loopEnacted = LoopEnacted(rate: tempBasal.value, duration: duration, timestamp: tempBasal.startDate, received:
+                true)
+            lastTempBasalUploaded = tempBasal
+        } else if let recommendation = recommendation {
+            // notEnacted
+            loopEnacted = LoopEnacted(rate: recommendation.rate, duration: recommendation.duration, timestamp: recommendation.recommendedDate, received: false)
+        } else {
+            loopEnacted = nil
+        }
+
+        let failureReason: String?
+
+
+        if let lastLoopError = lastLoopError {
+            failureReason = String(lastLoopError)
+        } else {
+            failureReason = nil
+        }
+
+        let loopName = NSBundle.mainBundle().bundleDisplayName
```

Comment:
Can we move lines like this out of the LoopDataManager and into the receiver's method?